src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.cs

```diff
@@ -98,7 +98,7 @@ internal JsonArray (JsonElement element, JsonNodeOptions? options = null) : base
         /// <param name="value">
         ///   The object to be added to the end of the <see cref="JsonArray"/>.
         /// </param>
-        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresUnreferencedCode(JsonValue.CreateUnreferencedCodeMessage)]
```

Comment:
Due to trimming this was discussed as a method that we may want to remove, and require usage of the `IList<T>.Add<T>(T value)` implementation. However since that requires an extra call to `JsonValue.Create()`, it affects usability.

Have there been other non-JSON areas that we've discouraged API use from in this manner in favor of forcing or encouraging a trimmable API?

Reply:
> Have there been other non-JSON areas that we've discouraged API use from in this manner in favor of forcing or encouraging a trimmable API?

Yes. A main difference between JSON and those APIs is that we are actively adding features to JSON, so that's why it seems JSON is trying to be "trim-compatible" from the start. And also because any API we've already shipped we can't change. But since we are defining these JSON APIs at the same time as making the BCL trimmable, we might as well at least consider it in the API design.

Other examples include:

* System.Linq.Expressions (ex. https://github.com/mono/linker/issues/1821 - see the full discussion, but the current recommended approach is to add a new overload that doesn't do the scanning for operator methods (which is the trim-incompatible issue)
* `TypeDescriptor` is virtually all `RequiresUnreferencedCode`, because of its heavy use of Reflection.
* `System.Data.Common` is getting a lot of `RequiresUnreferencedCode` attributes in https://github.com/dotnet/runtime/pull/52046

> discouraged API use

Note though, this only matters if you care about trimmability and AOT'ing your code. If a developer is only concerned about the "current deployment models" of their code, they can easily use these APIs and their code will work. They only get warnings once they attempt to trim unused code away.